notebooks/analyze_looping_across_lending_protocols.ipynb

```diff
@@ -373,32 +365,13 @@
     "                    user_actions[user].append(('debt', debt_token, debt_amount, protocol))\n",
     "    return user_actions\n",
     "    \n",
+    "    \n",
     "# Extract user actions\n",
     "user_actions = extract_user_actions(dataframes, tokens)\n",
     "\n",
-    "# Helper function to check if a loop exists and calculate total dollar amounts\n",
-    "def has_completed_loop(actions):\n",
-    "    token_graph = defaultdict(set)\n",
-    "    total_amount = 0.0\n",
-    "    collateral_amounts = defaultdict(float)\n",
-    "    debt_amounts = defaultdict(float)\n",
-    "    \n",
-    "    for action_type, token, amount, protocol in actions:\n",
-    "        if action_type == 'collateral':\n",
-    "            token_graph['collateral'].add(token)\n",
-    "            collateral_amounts[token] += amount\n",
-    "        elif action_type == 'debt':\n",
-    "            token_graph['debt'].add(token)\n",
-    "            debt_amounts[token] += amount\n",
-    "    \n",
-    "    for token in token_graph['collateral']:\n",
-    "        if token in token_graph['debt']:\n",
-    "            total_amount += min(collateral_amounts[token], debt_amounts[token])\n",
-    "    \n",
-    "    return total_amount\n",
     "\n",
-    "# Calculate total dollar amounts and looping users\n",
-    "def count_looping_users(user_actions):\n",
+    "# Calculate total looping users\n",
+    "def count_looping_users(user_actions: defaultdict[str, List[Tuple[str, str, float, str]]]) -> int:\n",
     "    looping_users = 0\n",
     "    for user, actions in user_actions.items():\n",
     "        loop_amount = has_completed_loop(actions)\n",
```

Comment:
Here you are using `has_completed_loop` before it's defined.

Reply:
Done! ✅